### ISSUE TITLE
Doesn't properly release VTE, causing mem and FD leak

### ISSUE THREAD

**Issue by egmontkob** [gnunn1/terminix#589]
Terminix and Terminator suffer from this same bug, so it's pretty much a copy-paste of https://bugs.launchpad.net/terminator/+bug/1645500. Other VTE-based terminals I've tried (gnome-terminal, xfce4-terminal, mate-terminal) don't have this issue.

---

As one uses Terminix, it opens up to 3 files (and immediately unlinks them) per VTE. These files contain the scrollback data.

When a terminal is closed, VTE is actually not properly released (apparently someone's still holding a reference [perhaps a callback, signal, whatever...]), and in turn its file descriptors are not closed. That is:

- All the disk space used by them to store the (potentially unlimited) scrollback remains occupied for the entire lifetime of Terminix, potentially leading to quickly filling up /tmp;

- There's a limit on the open FDs a process can have, it becomes possible that a user with a long-running Terminix hits this limit.

---

How to reproduce:

- Start Terminix with a couple of terminals;

- In each of them produce lot of output including various colors;

- ls -l /proc/[terminix's pid]/fd;

- Notice 3 file descriptors for each VTE pointing to "/tmp/blahblah (deleted)";

- Close some of the terminals;

- Notice that their files are still open.

- If you really produce a lot (like, really a lot) of data in these terminals with infinite scrollback, you can also see "df" reporting a shrinking /tmp.

- If you add a debug printf to vte's vte_terminal_finalize(), you'll see that it's not called.

Expected behavior:

- When a terminal is closed, its 3 file descriptors should disappear from this listing (at any time, the number of such deleted open files should be at most 3 times the number of open terminals). The available disk space ("df") under /tmp should grow back.

- The debug printf should tell you that vte_terminal_finalize() is executed.

---

A bit of technical background:

Each VTE creates 3 streams for the scrollback contents. One is for the plain text. The other is for colors and other attributes, using a runlength encoding (to--from coordinates for each continuous run of identical attributes). The third one contains pointers to the first two for each phyisical line of the buffer.

Each stream is flushed to disk after every 64kB of data. The file is only created at the first write, that is, when 64kB is already written to that stream.

There's compression, encryption, and a weird mapping to get a circular buffer if the scrollback is finite large, I'm not going into these details here.

If the teminal's contents consists of plain non-colored single-wide characters only, the file containing the attributes won't grow and hence this stream won't hit the disk. Therefore if you wish to see 3 file descriptors appearing, you need to produce various colors, or double-wide (CJK) characters; a giant plain English text will only get you 2 files.

When the terminal's contents are rewrapped on a resize, 1 of the 3 streams (the one with the row indexes) is regenerated and the old one is thrown away (and so is the underlying file). Therefore when a terminal is resized, you might face the number of opened files decreasing by 1.

I'm writing these so that you don't get into too many unknowns and wtfs while debugging :)

**Comment by gnunn1** [comment]
Gnome terminal is doing something with fd that it sets up manually, my C foo is fairly weak so I'm going to need some time to decipher it. But if you look here you can see where it sets it up:

https://github.com/GNOME/gnome-terminal/blob/master/src/terminal-screen.c#L1279

and on the spawn_async where it sets up a free function when it is destroyed:

https://github.com/GNOME/gnome-terminal/blob/master/src/terminal-screen.c#L1445

**Comment by gnunn1** [comment]
Although xfce is not doing anything fancy so it must be something else, will look into it further.

**Comment by egmontkob** [comment]
That's a different file descriptor. The ones I'm referring to are VTE's internal business, not exported towards the container app via any API or whatsoever. (And we've changed VTE's behavior around these quite a few times without the container apps noticing it or the API changing in any way.)

These file descriptors are just the easiest to see symptoms of the widget not being released. Another symptom is memory leak: the memory footprint of the entire app grows by perhaps another 0.5MB or so after each terminal that's opened, being used for a while and then closed.

---

A bit of technical background, again:

For each of these three streams, there's a 64kB write buffer and a 64kB read cache, that's 384kB. Plus all the other stuff, e.g. the onscreen contents, the input parsing (with a big ooops to gnome bug 730226 -- irrelevant now), all the internal structures and even the kitchen sink. Hence my wild guess of about half a meg of footprint per VTE.

**Comment by gnunn1** [comment]
As referenced above, I've opened an issue with GtkD along with a reproducer that demonstrates what I hope is causing the issue. Sometimes these things are multi-facted though and fixing this might just get me to the next round of debugging. Anyway, let's give the GtkD maintainer a few days to have a look at it and go from there.

**Comment by gnunn1** [comment]
The GtkD maintainer responded, while that does look like a leak the workaround I'm using of explicitly calling destroy seems to work fine, at least in my reproducer. I'm getting a bit stumped, I enhanced my reproducer to use the VTE widget and of course everything works perfectly including cleaning up the fd.

**Comment by gnunn1** [comment]
Did some thinking last night and made some progress before work this morning in terms of understanding the possible issues. I also found a potential bug in GtkD that prevents an object from being de-referenced to 0 if the destroy() method is called which I'm doing here and have updated the GtkD issue accordingly.

**Comment by gnunn1** [comment]
I've checked in a workaround that fixes the issue by simply removing the VTE widget manually from it's container and setting all references to null. While this fixes the VTE, there is still the issue of the Terminal object in terminal.d not being freed when removed from the session.

The terminal object refCount is 1 so it's not a case of GTK holding a reference somewhere, either there is a reference somewhere in my D code I am missing (most likely) or a problem with D's GC (less likely). More investigation required.

**Comment by gnunn1** [comment]
Made some good progress on this tonight, I'm not 100% sure but my style of handling events is causing a dangling reference to the delegates. I've been meaning to switch over to D's std.Signals mechanism for awhile this gave me an excuse to do so. With this change the Session object is being GC'ed correctly, still having an issue with the terminix Terminal (not VTE) object though but I'll look into it some more tomorrow.

**Comment by gnunn1** [comment]
OK this seems to be fixed now, as is typical with GC and reference issues there were several things that needed to be cleaned up:

1. The remove function in gx.util.array needs to set the element to null before removing it, this is because in D remove simply generates a slice pointing to the original storage. Since the original allocation still had a reference to the object it never got GC'ed.

2. Switched to D's signals/slots, suspect the problem with my original implementation was actually related to point 1 but I did notice in a reproducer it still seemed to have some issues. Since I had been planning on re-factoring to D's signals for awhile was a good excuse to do it.

3.Replaced a cross referenced event handler between terminal.d and search.d with a D signal since that supports removing event handlers

4. Aggressively removed the VTE widget from the Terminal so it always gets collected in case I screw this up again in the future

5. Using std.experimental.logger in destructors or Widget.addonDestroy handlers causes memory issues, removed in favor of writeln embedded in conditional debug block.

6. Added a finalizer to terminal.d so things can be cleaned up outside of the widgets destroy mechanism, useful for point 4.

Thanks for opening this issue @egmontkob , a total pain in the ass to fix as most memory issues are but well worth the effort.

Note for anyone testing this, D is a GC language thus those file descriptors will not disappear right away. In debug mode, the hamburger menu has a GC menu item that forces a collection and it often takes 3 to 4 clicks before everything gets cleaned up as it goes down through the layers.

**Comment by egmontkob** [comment]
Thanks a lot for this amazing investigation!

In the mean time I'm wondering if there's anything we could/should do in VTE to improve the situation (including getting rid of the GC delay).

Gtk+'s documentation is quite clear that closing FDs should happen in finalize(), not even in dispose() or similar. Especially with GC I'm not sure why. Maybe we should ask Gtk+ devs for best practice.

We could kinda go against this and have an optional explicit method to prematurely close the FDs and free up most of the memory. This could be a designated method, or we could hijack e.g. the existing clear-scrollback.

I can't remember if calling clear-scrollback from the API could free up the big amount of memory (that is, whether the memory is allocated on widget creation or on demand), but it could definitely close the FDs since there's nothing to be stored there at that point and the FDs are definitely created on demand.

We'd then need to differentiate between escape sequence initiated clear-scrollback (disputed anyways in [Gnome 771953](https://bugzilla.gnome.org/show_bug.cgi?id=771953)), and API initiated clear-scrollback. An escape sequence initiated one deliberately does not close the FDs and along with it definitely keeps the old encryption keys so that it can remain a fast operation (does not drain urandom etc. in case of malicious DoS-ing output contains many such sequences). We should definitely keep this. The API initiated one, on the other hand, would need to close the FDs and I can't remember if (without larger code modifications) this implies generating new encryption keys or not.

**Comment by egmontkob** [comment]
There's a security side of the story as well. The encryption key is destroyed when the VTE widget is finalized. That is, this is the moment from which someone getting access to the entire snapshot (RAM and disks) of the computer has practically no chance of recovering the scrollback data. I think users expect to happen this as soon as they close the terminal.

**Comment by gnunn1** [comment]
In general in languages with non-deterministic GC it is very common for resources to have some sort of ```close``` method on resources. My background is in Java and you get pretty used to closing resources like JDBC connections manually. So my two cents is yes, this would be very helpful. 

The other option is that most GC languages allow the developer to force a collection, I tried that here but like I said it takes 2-4 collections before it all get's cleaned up in D so it is sub-optimal for me.

I'd very interested in understanding how Python is handling this versus D, hopefully the terminator maintainer will have a look at this issue in the near future.

**Comment by sboddy** [comment]
Terminator maintainer here. Just an FYI here, so feel free to ignore it. Finally got to the bottom of this one. For us this bug was first raised waaaaay back in 2011, but it was never on my radar (I became maintainer after this and we have a lot of old crufty out of date bugs). Then Egmont raised a new issue with all the excellent detail, and this bug was useful for clues to the fix, so many thanks for doing the initial heavy lifting. For us the fix was in four parts, and it seemed all were required.

1. We had to remove the vte widget from the container terminal object (i.e. self.terminalbox.remove(self.vte)),
2. We were already using a signal handler/record, but only for a few of the numerous signals we connect to, so I changed all the other signals to setup using the handler,
3. We had to remove all the signal handlers, rather than just one, so the remove was changed to remove all signals at once, rather than individually,
4. Had to account for two slightly different code paths doing inconsistent things.

After those four changes, the fd's were properly released. Like for you, it was a pain in the butt to fix. I'm not sure if I can shed any light on the internal Python workings, beyond saying that Python used refcounts and gc, but I didn't have to force anything (i.e. garbage collections) after the above changes. The fd's are released immediately.

Here's the fix: http://bazaar.launchpad.net/~gnome-terminator/terminator/gtk3/revision/1729

The truth is, I'm not what I consider a "real" developer, so I fumble about quite a bit and half of the time it's trail and error till I find a fix. I expect there's probably quite a few other similarly leaky corners to the Terminator codebase. I did however learn quite a bit doing this one :-)
